Manual: Add more context to the description of locally abstract types

Octachron opened the pull request:
This minor pull request adds a little more context to the description of locally abstract types.
More precisely, this PR 
- add two forward reference to the sections on GADTs and first-class modules cited in this description
- use the full name of GADTs when first referencing them
- mention that the polymorphic syntax is particularly useful when defining polymorphically recursive functions involving GADTs.

Octachron committed:
Manual: polymorphic abstract types: more context

gasche commented:
I like the idea of being explicit about polymorphic recursion plus
gadts, but I think we could do better. First, "polymorphically
recursive" is a scary word (I think "polymorphic recursion" should be
the limit of language complexity here), and more importantly this
comment does not provide a lot of guidance for users (it "can" be
useful, but why, when is it necessary?). I think we could move this
mixed syntax to after the remark on `method!`, be more explanatory,
and give an example. Attempt below.

> The `typeconstr-name` syntax construction by itself does not make
> polymorphic the type variable it introduces, but it can be combined
> with explicit polymorphic annotations where needed.  The above rule is
> provided as syntactic sugar to make this easier:
> 
> ```
> let rec f : type t1 t2. t1 * t2 list -> t1 = ...
> ```
> 
> is automatically expanded into
> 
> ```
> let rec f : 't1 't2. 't1 * 't2 list -> 't1 =
>   fun (type t1) (type t2) -> (... : t1 * t2 list -> t1)
> ```
> 
> When defining a recursive function that manipulates a GADT, you
> should generally use explicit polymorphic recursion. For example,
> the following works:
> 
> ```
> type _ expr =
>   | Int : int -> int expr
>   | Pair : 'a expr * 'b expr -> ('a * 'b) expr
> 
> let rec eval : type a . a expr -> a = function
>   | Int n -> n
>   | Pair (a, b) -> (eval a, eval b)
> ```
> 
> but the following results in a type error
> 
> ```
> let rec eval (type a) : a expr -> a = function
>   | Int n -> n
>   | Pair (a, b) -> (eval a, eval b)
> (*                       ^
> Error: This expression has type $0 expr
>        but an expression was expected of type 'a
>        The type constructor $0 would escape its scope
>  *)
> ```
> 
> In absence of an explicit polymorphic annotation, a monomorphic
> type is inferred for the recursive function. If a recursive call inside the
> definition occurs at a type that involves an existential GADT
> variable, this variable flows to the type of the recursive function,
> and thus escapes its scope.

Octachron commented:
I like the idea of having a detailed example. However, I am not sure that the locally abstract type section is the right place for this example: I would like to preserve the fiction that the manual can be read linearly. That's why, I think that it would be better to explain the motivation behind the `type a.` annotation with the first example in the GADTs section rather than add a GADT example inside the locally abstract type section. The locally abstract section could then be updated with just a "see the GADT section for a more detailed explanation". (And I woud replace the scary "polymorphically".)

gasche commented:
That sounds very good -- you have a better picture of the documentation than I do. Should I merge the current PR as-is, then?

Drup commented:
I still think @gasche's paragraph is quite good and should be included (not sure where, though).

Octachron committed:
Manual: avoid "polymorphically"

Octachron committed:
Manual: detail polymorphic annotation for gadts

Octachron commented:
@gasche, Well,  while fixing the scary "polymorphically", I ended up adding your proposed explanation.

@Drup, I agree. 

I have added slightly reworked version of this paragraph in the GADTs section. It still needs some work however (it should at least mention that $ means existential type variable).

Octachron committed:
Manual: gadt monomorphic error explanation

This commit rewords the explanation of the erroneously monomorphic
example in the GADT section and add a forward reference to the naming
convention followed by the compiler for existential type appearing in
error messages.

Octachron commented:
I have updated the explanation on the monomorphic GADT example to refer to the paragraph on the existential type nomenclature.

gasche commented:
Very nice! Maybe we could add two subsections to the GADT section of the manual, one "recursive functions on GADTs", and one "GADTs and type inference"? Otherwise I feel that this version is ready to merge.

Octachron committed:
Manual: paragraph headers in section 7.16 (GADTs)

Octachron commented:
The GADTs section is currently using paragraph title rather than subsection. To minimize changes, I have thus added new paragraph titles. Another option would be to promote all these paragraph titles to subsection. I have also not included GADTs in these titles, since the term is already repeated quite often within the section.

gasche commented:
Thanks for the patch. I merged it in trunk, and will also commit in 4.03.

gasche commented:
(While we're discussing documentation, let me apologize for not having found the time to work on those warning explanations yet. I would still like to work on this, but I've been working on my thesis (and now defense) in the past few months.)

Octachron commented:
( There is no need to apologize, I am well aware that academic time can be quite ... fluctuating. Good luck with your defense.)
